### ISSUE TITLE
Can we support creating `Evaluator` objects from strings?

### ISSUE THREAD

**Issue by marlonlcz** [jpmml/jpmml-evaluator-python#28]
When I use "jpmml-evaluator-python", I want to create an "evaluator" object directly from a python string. 
However, the existing function can only support creating this object from a pmml file.

**Comment by vruusmann** [comment]
> I want to create an "evaluator" object directly from a python string.

Where does this Python string come from? Is it generated on-the-fly (by some Python function), or is it loaded from some external non-file resource (eg. downloading by URL)?

If it's the latter, then perhaps the real feature request should be "load PMML from non-file resources".

> However, the existing function can only support creating this object from a pmml file.

The `LoadingModelEvaluatorBuilder` class currently defines a `loadFile(path)` method:
https://github.com/jpmml/jpmml-evaluator-python/blob/0.10.5/jpmml_evaluator/__init__.py#L230-L233

For direct loading from strings, we would simply need to add a `loadString(pmml_str)` method there. However, this Python method would need to have a backing Java method in the core [JPMML-Evaluator](https://github.com/jpmml/jpmml-evaluator) library.

There are some quick workarounds available, though. For example, the `loadString(pmml_str)` might dump the string to a temporary file, and call `loadFile(path)` on it.

But still, would love to hear more about the origin of PMML strings first.

**Comment by marlonlcz** [comment]
PMML string created using Python function.

In the Java version, there is a LoadingModelEvaluatorBuilder class that can support inputting string model PMML files. 
Can this method be added to the Python package?

**Comment by marlonlcz** [comment]
> > I want to create an "evaluator" object directly from a python string.
> 
> Where does this Python string come from? Is it generated on-the-fly (by some Python function), or is it loaded from some external non-file resource (eg. downloading by URL)?
> 
> If it's the latter, then perhaps the real feature request should be "load PMML from non-file resources".
> 
> > However, the existing function can only support creating this object from a pmml file.
> 
> The `LoadingModelEvaluatorBuilder` class currently defines a `loadFile(path)` method: https://github.com/jpmml/jpmml-evaluator-python/blob/0.10.5/jpmml_evaluator/__init__.py#L230-L233
> 
> For direct loading from strings, we would simply need to add a `loadString(pmml_str)` method there. However, this Python method would need to have a backing Java method in the core [JPMML-Evaluator](https://github.com/jpmml/jpmml-evaluator) library.
> 
> There are some quick workarounds available, though. For example, the `loadString(pmml_str)` might dump the string to a temporary file, and call `loadFile(path)` on it.
> 
> But still, would love to hear more about the origin of PMML strings first.

In the Java version, there is a LoadingModelEvaluatorBuilder class that can support inputting string model PMML files. Can this method be added to the Python package?

**Comment by lchzh** [comment]
I made a simple rewrite of the python version to load PMML from a string. If possible, please add it to the release version.

class LoadingModelEvaluatorBuilder(BaseModelEvaluatorBuilder):
        ......
	def LoadStr(self, s):
		model = self.backend.newObject("java.lang.String", s)
		model_stream = self.backend.newObject("java.io.ByteArrayInputStream", model.getBytes())
		self.javaModelEvaluatorBuilder.load(model_stream)
		return self

def make_evaluator(obj, backend = "jpype", lax = False, locatable = False, reporting = False, transpile = False):
	""" Builds an Evaluator based on a PMML file.

	Parameters:
	----------
	obj: string/bytes
		The path to the PMML file in local filesystem, a string(pmml), or an array of bytes(bytes or bytearray)

	backend: JavaBackend or string
		The Java backend or its alias

	lax: boolean
		If True, skip model schema sanity checks.

	locatable: boolean
		If True, retain SAX Locator information (if available),
		which leads to more informative exception messages.

	reporting: boolean
		If True, activate the reporting Value API.

	transpile: boolean or string
		If not False, perform transpilation.
	"""

	if isinstance(backend, JavaBackend):
		pass
	elif isinstance(backend, str):
		backend = make_backend(backend)
	else:
		raise TypeError()
	model_obj = obj
	if hasattr(obj, 'read') and callable(obj.read):
		model_obj = obj.read()

	if isinstance(model_obj, (bytes, bytearray)):
		model_obj = model_obj.decode('utf-8')

	if isinstance(model_obj, (str, u"".__class__)):
		if os.path.exists(model_obj):
			evaluatorBuilder = LoadingModelEvaluatorBuilder(backend, lax) \
				.setLocatable(locatable) \
				.loadFile(model_obj)
		else:
			evaluatorBuilder = LoadingModelEvaluatorBuilder(backend, lax) \
				.setLocatable(locatable) \
				.loadStr(model_obj)
	else:
		raise ValueError('Input type "{type}" not supported'.format(type=type(model_obj).__name__))

	if reporting:
		evaluatorBuilder.setReportingValueFactoryFactory()
	if transpile:
		evaluatorBuilder.transpile(transpile if isinstance(transpile, str) else None)
	return evaluatorBuilder.build()

**Comment by marlonlcz** [comment]
I made a simple rewrite of the python version to load PMML from a string, if possible, please add it to the release.

"""

class LoadingModelEvaluatorBuilder(BaseModelEvaluatorBuilder):
         .......
	def LoadStr(self, s):
		model = self.backend.newObject("java.lang.String", s)
		model_stream = self.backend.newObject("java.io.ByteArrayInputStream", model.getBytes())
		self.javaModelEvaluatorBuilder.load(model_stream)

def make_evaluator(obj, backend = "jpype", lax = False, locatable = False, reporting = False, transpile = False):
	""" Builds an Evaluator based on a PMML file.

	Parameters:
	----------
	obj: string/bytes
		The path to the PMML file in local filesystem, a string(pmml), or an array of bytes(bytes or bytearray)

	backend: JavaBackend or string
		The Java backend or its alias

	lax: boolean
		If True, skip model schema sanity checks.

	locatable: boolean
		If True, retain SAX Locator information (if available),
		which leads to more informative exception messages.

	reporting: boolean
		If True, activate the reporting Value API.

	transpile: boolean or string
		If not False, perform transpilation.
	"""

	if isinstance(backend, JavaBackend):
		pass
	elif isinstance(backend, str):
		backend = make_backend(backend)
	else:
		raise TypeError()
	model_obj = obj
	if hasattr(obj, 'read') and callable(obj.read):
		model_obj = obj.read()

	if isinstance(model_obj, (bytes, bytearray)):
		model_obj = model_obj.decode('utf-8')

	if isinstance(model_obj, (str, u"".__class__)):
		if os.path.exists(model_obj):
			evaluatorBuilder = LoadingModelEvaluatorBuilder(backend, lax) \
				.setLocatable(locatable) \
				.loadFile(model_obj)
		else:
			evaluatorBuilder = LoadingModelEvaluatorBuilder(backend, lax) \
				.setLocatable(locatable) \
				.loadStr(model_obj)
	else:
		raise ValueError('Input type "{type}" not supported'.format(type=type(model_obj).__name__))

	if reporting:
		evaluatorBuilder.setReportingValueFactoryFactory()
	if transpile:
		evaluatorBuilder.transpile(transpile if isinstance(transpile, str) else None)
	return evaluatorBuilder.build()

**Comment by marlonlcz** [comment]
I made a simple rewrite of the python version to load PMML from a string, if possible, please add it to the release.

`class LoadingModelEvaluatorBuilder(BaseModelEvaluatorBuilder):

	def __init__(self, backend, lax = False):
		javaModelEvaluatorBuilder = backend.newObject("org.jpmml.evaluator.LoadingModelEvaluatorBuilder")
		super(LoadingModelEvaluatorBuilder, self).__init__(backend, javaModelEvaluatorBuilder)
		self.setCheckSchema(not lax)

	def setLocatable(self, locatable = False):
		self.javaModelEvaluatorBuilder.setLocatable(locatable)
		return self

	def loadFile(self, path):
		file = self.backend.newObject("java.io.File", path)
		self.javaModelEvaluatorBuilder.load(file)
		return self

	def LoadStr(self, s):
		model = self.backend.newObject("java.lang.String", s)
		model_stream = self.backend.newObject("java.io.ByteArrayInputStream", model.getBytes())
		self.javaModelEvaluatorBuilder.load(model_stream)
		return self

	def transform(self, javaPMMLTransformer):
		self.javaModelEvaluatorBuilder.transform(javaPMMLTransformer)
		return self

	def transpile(self, path = None):
		if path:
			file = self.backend.newObject("java.io.File", path)
			transpiler = self.backend.newObject("org.jpmml.transpiler.FileTranspiler", None, file)
		else:
			transpiler = self.backend.newObject("org.jpmml.transpiler.InMemoryTranspiler", None)
		transformer = self.backend.newObject("org.jpmml.transpiler.TranspilerTransformer", transpiler)
		return self.transform(transformer)

def make_evaluator(obj, backend = "jpype", lax = False, locatable = False, reporting = False, transpile = False):
	""" Builds an Evaluator based on a PMML file.

	Parameters:
	----------
	obj: string/bytes
		The path to the PMML file in local filesystem, a string(pmml), or an array of bytes(bytes or bytearray)

	backend: JavaBackend or string
		The Java backend or its alias

	lax: boolean
		If True, skip model schema sanity checks.

	locatable: boolean
		If True, retain SAX Locator information (if available),
		which leads to more informative exception messages.

	reporting: boolean
		If True, activate the reporting Value API.

	transpile: boolean or string
		If not False, perform transpilation.
	"""

	if isinstance(backend, JavaBackend):
		pass
	elif isinstance(backend, str):
		backend = make_backend(backend)
	else:
		raise TypeError()
	model_obj = obj
	if hasattr(obj, 'read') and callable(obj.read):
		model_obj = obj.read()

	if isinstance(model_obj, (bytes, bytearray)):
		model_obj = model_obj.decode('utf-8')

	if isinstance(model_obj, (str, u"".__class__)):
		if os.path.exists(model_obj):
			evaluatorBuilder = LoadingModelEvaluatorBuilder(backend, lax) \
				.setLocatable(locatable) \
				.loadFile(model_obj)
		else:
			evaluatorBuilder = LoadingModelEvaluatorBuilder(backend, lax) \
				.setLocatable(locatable) \
				.loadStr(model_obj)
	else:
		raise ValueError('Input type "{type}" not supported'.format(type=type(model_obj).__name__))

	if reporting:
		evaluatorBuilder.setReportingValueFactoryFactory()
	if transpile:
		evaluatorBuilder.transpile(transpile if isinstance(transpile, str) else None)
	return evaluatorBuilder.build()`

**Comment by marlonlcz** [comment]
I made a simple rewrite of the python version to load PMML from a string, if possible, please add it to the release.

```python
class LoadingModelEvaluatorBuilder(BaseModelEvaluatorBuilder):

	def __init__(self, backend, lax = False):
		javaModelEvaluatorBuilder = backend.newObject("org.jpmml.evaluator.LoadingModelEvaluatorBuilder")
		super(LoadingModelEvaluatorBuilder, self).__init__(backend, javaModelEvaluatorBuilder)
		self.setCheckSchema(not lax)

	def setLocatable(self, locatable = False):
		self.javaModelEvaluatorBuilder.setLocatable(locatable)
		return self

	def loadFile(self, path):
		file = self.backend.newObject("java.io.File", path)
		self.javaModelEvaluatorBuilder.load(file)
		return self

	def LoadStr(self, s):
		model = self.backend.newObject("java.lang.String", s)
		model_stream = self.backend.newObject("java.io.ByteArrayInputStream", model.getBytes())
		self.javaModelEvaluatorBuilder.load(model_stream)
		return self

	def transform(self, javaPMMLTransformer):
		self.javaModelEvaluatorBuilder.transform(javaPMMLTransformer)
		return self

	def transpile(self, path = None):
		if path:
			file = self.backend.newObject("java.io.File", path)
			transpiler = self.backend.newObject("org.jpmml.transpiler.FileTranspiler", None, file)
		else:
			transpiler = self.backend.newObject("org.jpmml.transpiler.InMemoryTranspiler", None)
		transformer = self.backend.newObject("org.jpmml.transpiler.TranspilerTransformer", transpiler)
		return self.transform(transformer)
```

```python
def make_evaluator(obj, backend = "jpype", lax = False, locatable = False, reporting = False, transpile = False):

	if isinstance(backend, JavaBackend):
		pass
	elif isinstance(backend, str):
		backend = make_backend(backend)
	else:
		raise TypeError()
	model_obj = obj
	if hasattr(obj, 'read') and callable(obj.read):
		model_obj = obj.read()

	if isinstance(model_obj, (bytes, bytearray)):
		model_obj = model_obj.decode('utf-8')

	if isinstance(model_obj, (str, u"".__class__)):
		if os.path.exists(model_obj):
			evaluatorBuilder = LoadingModelEvaluatorBuilder(backend, lax) \
				.setLocatable(locatable) \
				.loadFile(model_obj)
		else:
			evaluatorBuilder = LoadingModelEvaluatorBuilder(backend, lax) \
				.setLocatable(locatable) \
				.loadStr(model_obj)
	else:
		raise ValueError('Input type "{type}" not supported'.format(type=type(model_obj).__name__))

	if reporting:
		evaluatorBuilder.setReportingValueFactoryFactory()
	if transpile:
		evaluatorBuilder.transpile(transpile if isinstance(transpile, str) else None)
	return evaluatorBuilder.build()
```

**Comment by vruusmann** [comment]
Noted.

Yes, it is a much better idea to transform the PMML string first to a byte array, and then to `java.io.ByteArrayInputStream`, and do the loading using the existing `LoadingModelEvaluatorBuilder#load(InputStream)` method. There is no need to dump the string to a temporary file - at least when the string isn't too long.

Also, I can see that you want to have the "load from string" functionality available via the `make_evaluator(...)` utility function.

Give me a couple of days.

**Comment by vruusmann** [comment]
The update has been released to PyPI as the 0.11.1 version:
https://pypi.org/project/jpmml-evaluator/0.11.1/

To summarize, the `jpmml_evaluator.make_evaluator()` utility function now accepts `str`, `bytes` and `pathlib.Path` objects as the main argument:
https://github.com/jpmml/jpmml-evaluator-python/blob/0.11.1/jpmml_evaluator/__init__.py#L313-L323

**Comment by marlonlcz** [comment]
> The update has been released to PyPI as the 0.11.1 version: https://pypi.org/project/jpmml-evaluator/0.11.1/
> 
> To summarize, the `jpmml_evaluator.make_evaluator()` utility function now accepts `str`, `bytes` and `pathlib.Path` objects as the main argument: https://github.com/jpmml/jpmml-evaluator-python/blob/0.11.1/jpmml_evaluator/__init__.py#L313-L323

Well done! Thank you for the fix and update! I will try to use the latest version of python package.
